### ISSUE TITLE
Detections path

### ISSUE THREAD

**Issue by luckylukez** [orrzohar/PROB#58]
Hi Orr, 
Thanks for great work!

I would like to save the detections done during validation. Is there any existing functionality for this or do I need to implement it?

**Comment by orrzohar** [comment]
Hi @luckylukez,
Do you mean the bbox/classes predicted?
If you do, then no I don't have this functionality. However, it should be really easy to add.
`results` here:
https://github.com/orrzohar/PROB/blob/28afbc1d7f5bdfb9a384ce09c870fcd829adc1b2/engine.py#L122

is either a list of dicts (corresponding to the different images in the batch), with the keys being the scores/labels/bboxs and values being the lists of the detections. All you need to do is add before the main for loop:

`all_det=[]`

Then, in line 123:

`all_det+=results`

You then can either return `all_det` and save it in the function that calls `evaluate` or in `evaluate` itself as a json file. I would include the `epoch` if this is during training.

Here is how it could potentially look:


```code
@torch.no_grad()
def evaluate(model, criterion, postprocessors, data_loader, base_ds, device, output_dir, args):
    model.eval()
    criterion.eval()
    metric_logger = utils.MetricLogger(delimiter="  ")
    header = 'Test:'
    iou_types = tuple(k for k in ('segm', 'bbox') if k in postprocessors.keys())
    coco_evaluator = OWEvaluator(base_ds, iou_types, args=args)
 
    panoptic_evaluator = None
    if 'panoptic' in postprocessors.keys():
        panoptic_evaluator = PanopticEvaluator(
            data_loader.dataset.ann_file,
            data_loader.dataset.ann_folder,
            output_dir=os.path.join(output_dir, "panoptic_eval"),
        )
 
    all_det = [] ## edit 1
    for samples, targets in metric_logger.log_every(data_loader, 10, header):
        samples = samples.to(device)
        targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
        outputs = model(samples)

        orig_target_sizes = torch.stack([t["orig_size"] for t in targets], dim=0)
        results = postprocessors['bbox'](outputs, orig_target_sizes)
        all_det += results ## edit 2
 
        if 'segm' in postprocessors.keys():
            target_sizes = torch.stack([t["size"] for t in targets], dim=0)
            results = postprocessors['segm'](results, outputs, orig_target_sizes, target_sizes)
        res = {target['image_id'].item(): output for target, output in zip(targets, results)}
        if coco_evaluator is not None:
            coco_evaluator.update(res)
 
        if panoptic_evaluator is not None:
            res_pano = postprocessors["panoptic"](outputs, target_sizes, orig_target_sizes)
            for i, target in enumerate(targets):
                image_id = target["image_id"].item()
                file_name = f"{image_id:012d}.png"
                res_pano[i]["image_id"] = image_id
                res_pano[i]["file_name"] = file_name
 
            panoptic_evaluator.update(res_pano)
 
    # gather the stats from all processes
    metric_logger.synchronize_between_processes()
    # print("Averaged stats:", metric_logger)
    if coco_evaluator is not None:
        coco_evaluator.synchronize_between_processes()
    if panoptic_evaluator is not None:
        panoptic_evaluator.synchronize_between_processes()
 
    # accumulate predictions from all images
    if coco_evaluator is not None:
        coco_evaluator.accumulate()
        res = coco_evaluator.summarize()
    panoptic_res = None
    if panoptic_evaluator is not None:
        panoptic_res = panoptic_evaluator.summarize()
    stats = {k: meter.global_avg for k, meter in metric_logger.meters.items()}
    stats['metrics']=res
    if coco_evaluator is not None:
        if 'bbox' in postprocessors.keys():
            stats['coco_eval_bbox'] = coco_evaluator.coco_eval['bbox'].stats.tolist()
        if 'segm' in postprocessors.keys():
            stats['coco_eval_masks'] = coco_evaluator.coco_eval['segm'].stats.tolist()
    if panoptic_res is not None:
        stats['PQ_all'] = panoptic_res["All"]
        stats['PQ_th'] = panoptic_res["Things"]
        stats['PQ_st'] = panoptic_res["Stuff"]

    if args.save_eval_det_file: ## edit 3
        with open(args.save_eval_det_file, 'w') as f: ## edit 3
            json.dump(all_det, f, indent=4) ## edit 3
    return stats, coco_evaluator

```

Best,
Orr

**Comment by luckylukez** [comment]
Yes I meant the bbox and classes predicted.
Thank you very much for the helpful and timely answer!

**Comment by luckylukez** [comment]
Follow up:

Since the results are torch tensors I found that i first had to convert them to lists in order to save them as json. Before '## edit 3' I added this code:

```
all_det_jsonified = []
for item in all_det:
  converted_dict = {key: value.cpu().numpy().tolist() for key, value in item.items()}
  all_det_jsonified.append(converted_dict)
```
